core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -509,6 +511,13 @@ private[spark] class ExecutorAllocationManager(
   private def onExecutorBusy(executorId: String): Unit = synchronized {
     logDebug(s"Clearing idle timer for $executorId because it is now running a task")
     removeTimes.remove(executorId)
+
+    // Executor is added to remove by misjudgment due to async listener making it as idle).
+    // see SPARK-9552
+    if (executorsPendingToRemove.contains(executorId)) {
```

Comment:
@vanzin Here is the code path.

1. Prepare entire executorID-list to be killed (meet certain criteria) 
2. killExecutors will filter out non-eligible ones (some of them may not be killed accordingly)
3. no matter what kind of executors filtered out, if some of them are acknowledged(really killed), we will add all of the executorID-list to `executorsPendingToRemove`. There is no way to tell who is actually to kill. 

That is why we need such kind of rescuing. please let me know if it makes sense.